packages/loopring_v3/contracts/iface/ExchangeData.sol

```diff
@@ -155,66 +175,65 @@ library ExchangeData
         bool    onchainDataAvailability;
         bytes32 genesisMerkleRoot;
 
+        bytes32 DOMAIN_SEPARATOR;
+
         ILoopringV3      loopring;
         IBlockVerifier   blockVerifier;
         IDepositContract depositContract;
 
-        address lrcAddress;
-
         uint    totalTimeInMaintenanceSeconds;
         uint    numDowntimeMinutes;
         uint    downtimeStart;
 
-        address addressWhitelist;
-        uint    accountCreationFeeETH;
-        uint    accountUpdateFeeETH;
-        uint    depositFeeETH;
-        uint    withdrawalFeeETH;
+        // List of all tokens
+        Token[] tokens;
 
-        Token[]     tokens;
-        Account[]   accounts;
-        Deposit[]   deposits;
-        Request[]   depositChain;
-        Request[]   withdrawalChain;
+        // List of all blocks
+        BlockInfo[] blocks;
+
+        // A map from a token to its tokenID + 1
+        mapping (address => uint16) tokenToTokenId;
 
         // The merkle root of the offchain data stored in a Merkle tree. The Merkle tree
         // stores balances for users using an account model.
         bytes32 merkleRoot;
 
-        // The number of blocks that are submitted onchain
-        uint32 numBlocksSubmitted;
-
-        // The number of onchain deposit requests that have been processed
-        // up to and including this block.
-        uint32 numDepositRequestsCommitted;
+        // A map from an accountID to a tokenID to if the balance is withdrawn
+        mapping (uint24 => mapping (uint16 => bool)) withdrawnInWithdrawMode;
 
-        // The number of onchain withdrawal requests that have been processed
-        // up to and including this block.
-        uint32 numWithdrawalRequestsCommitted;
+        // A map from an account to a token to the amount withdrawable for that account.
+        // This is only used when the automatic distribution of the withdrawal failed.
+        mapping (address => mapping (uint16 => uint)) amountWithdrawable;
```

Comment:
Currently they will still be part of the total rollup balance, so index calculations would still distribute profits to them (which would indeed be lost).

There are definitely some scenarios now where there are some inefficiencies to keep things as simple as possible. Not a problem if the index can only increase, but definitely a problem when the index can also decrease.

> I think the deposit contract impl can enable these extra fund to be claimed by the developer somehow...

I think it's not really possible as of now (would have been possible with the complete shutdown mode where the complete Merkle tree needed to be reset, but we don't do that any more), unless we track everything 100% correctly (so even pass the balance index on-chain for withdrawals and still update balances there when necessary), in which case there would be no extra funds I guess (unless we choose not to give them). :)

Reply:
Brecht, are we assuming the index is ever-growing or it can also go down (implying a lost). I think we need to make the right assumption, very important.

My understanding of the current withdrawal behaviors:

- the relayer will calculate how much token the user actually has, this is the process of using the last index,  the current index, and the last balance to calculate the most up-to-date balance. If the balance >= amountToWithdraw, amountToWithdraw will be withdrawn.  From this moment on, the `amountToWithdrawl` part is no longer earning interest, positive or negative, for the user, although the corresponding token in DepositContract still earns interest.

- If amountToWithdraw is not claimed, and the DepositContract end up with a negative earning after the withdrawal, then in a rare situation where there is only one user, the DepositContract's total balance will be smaller than amountToWithdraw, the user will fail to claim the token.

Should we allow the user to claim as much as possible instead of a failure?